src/main/java/io/nats/client/impl/NatsMessage.java

```diff
@@ -13,195 +13,184 @@
+    private Kind kind;
+    private int sizeInBytes = -1;
+    private int hdrLen = 0;
+    private int dataLen = 0;
+    private int totLen = 0;
+
     private NatsSubscription subscription;
-    private Integer protocolLength = null;
 
     NatsMessage next; // for linked list
 
+    public NatsMessage(String subject, String replyTo, byte[] data, boolean utf8mode) {
+        this(subject, replyTo, null, data, utf8mode);
+    }
+
+    public NatsMessage(Message message) {
+        this(message.getSubject(), message.getReplyTo(),
+                message.getHeaders(), message.getData(), message.isUtf8mode());
+    }
+
     // Create a message to publish
-    NatsMessage(String subject, String replyTo, ByteBuffer data, boolean utf8mode) {
+    public NatsMessage(String subject, String replyTo, Headers headers, byte[] data, boolean utf8mode) {
+
+        if (subject == null || subject.length() == 0) {
+            throw new IllegalArgumentException("Subject is required");
+        }
+
+        if (replyTo != null && replyTo.length() == 0) {
+            throw new IllegalArgumentException("ReplyTo cannot be the empty string");
+        }
+
+        kind = Kind.REGULAR;
         this.subject = subject;
         this.replyTo = replyTo;
-        this.data = data;
-        Charset charset;
-
-        // Calculate the length in bytes
-        int size = (data != null) ? data.limit() : 0;
-        int len = fastIntLength(size) + 4;
-        if (replyTo != null) {
-            if (utf8mode) {
-                len += fastUtf8Length(replyTo) + 1;
-            } else {
-                len += replyTo.length() + 1;
-            }
+        this.headers = headers;
+        this.data = data == null ? EMPTY_BODY : data;
+        this.utf8mode = utf8mode;
+
+        int replyToLen = replyTo == null ? 0 : replyTo.length();
+        dataLen = this.data.length;
+        if (headers != null && !headers.isEmpty()) {
+            hdrLen = headers.serializedLength();
         }
-        if (utf8mode) {
-            len += fastUtf8Length(subject) + 1;
-            charset = StandardCharsets.UTF_8;
-        } else {
-            len += subject.length() + 1;
-            charset = StandardCharsets.US_ASCII;
+        else {
+            hdrLen = 0;
         }
-        this.protocolBytes = ByteBuffer.allocate(len);
-        protocolBytes.put((byte)'P').put((byte)'U').put((byte)'B').put((byte)' ');
-        protocolBytes.put(subject.getBytes(charset));
-        protocolBytes.put((byte)' ');
+        totLen = hdrLen + dataLen;
 
-        if (replyTo != null) {
-            protocolBytes.put(replyTo.getBytes(charset));
-            protocolBytes.put((byte)' ');
+        // initialize the builder with a reasonable length, preventing resize in 99.9% of the cases
+        // 32 for misc + subject length doubled in case of utf8 mode + replyToLen + totLen (hdrLen + dataLen)
+        ByteArrayBuilder bab = new ByteArrayBuilder(32 + (subject.length() * 2) + replyToLen + totLen);

```

Comment:
This has the potential to cause subtle breakage if you don't precisely allocate the correct buffer size.